packages/gasket-plugin-typescript/docs/esm-typescript.md

```diff
@@ -0,0 +1,77 @@
+# ESM and TypeScript
+
+## Introduction
+
+The introduction of ECMAScript Modules (ESM) in JavaScript has standardized how modules are handled, moving away from CommonJS. However, integrating ESM with TypeScript introduces several complications due to differences in how both systems handle modules, especially concerning file extensions, module resolution, and the interoperation between TypeScript and JavaScript files.
+
+## File Extensions in ESM
+
+In ESM, explicit file extensions are required when importing modules. This is a departure from CommonJS, where file extensions were often inferred, especially in TypeScript projects. For example, in ESM:
+
+```js
+import { myFunction } from './myModule.js';
+```
+
+In contrast, in TypeScript (when using CommonJS):
+
+```ts
+import { myFunction } from './myModule';
+```
+
+When switching to ESM, TypeScript does not automatically append file extensions, leading to runtime errors if the correct files are not found. To comply with ESM's requirements, developers must include the file extensions explicitly:
+
+```ts
+import { myFunction } from './myModule.js';
+```
```

Comment:
The rendered markdown does not surface if this is ts or js. Can you add comments to these blocks and the ones below?

> ```diff
> + //example.js
> import { myFunction } from './myModule.js';
> ```
> 
> In contrast, in TypeScript (when using CommonJS):
> 
> ```diff
> + // example.ts
> import { myFunction } from './myModule';
> ```
> 
> When switching to ESM, TypeScript does not automatically append file extensions, leading to runtime errors if the correct files are not found. To comply with ESM's requirements, developers must include the file extensions explicitly:
> 
> ```diff
> + // example.ts
> import { myFunction } from './myModule.js';
> ```